keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,676 @@
+    an error otherwise.  This means that sibling Kustomizations are entirely
+    independent of each other.
+  * Kustomizations with the same base will cause an error when the resources
+    are accumulated, unless their GVKN is altered.
+
+`Components` are processed between step 1 and 2, and in a similar manner to
+Kustomizations, but instead of starting with an empty RA, they take the RA from
+their parent.  The component is therefore able to modify everything within the
+RA, which will contain all of the parent's resources and the result of
+processing all of the earlier components in the `components:` list. Nested
+components will work in the same way - ownership of the parent's RA will simply
+be passed down.
+
+This was implemented by changing the signature of
+`kusttarget.accumulateResources` to return a pointer to the
+`ResourceAccumulator` that the parent should use, and giving child components
+their parent's RA instead of an empty RA. This minimized the amount of code
+changed to simplify review, but a slightly more extensive refactoring would be
+recommended.
+
+Additional code changes were required to ensure that Kustomizations and normal
+files are not added to the `components:` list and Components are not added to
+`resources:`.
+
+No changes were made to the code that loads the contents of Kustomizations, so
+it should be possible to load Components from any place where one expects to
+load a Kustomization.
+
+A variation of the implementation would be to allow Components to be added to
+the `resources:` list (instead of adding them to a new `components:` list):
+
+* The basic implementation of components wouldn't need to change.
+* Unless a naming convention would be used, it would be unclear whether
+  directories in the `resources:` list are Kustomizations or Components.  A
+  naming convention could be used, but this is less explicit than using a
+  separate `components:` list. For example, store components in a
+  `components/` directory in the same way as overlay Kustomizations are often
+  stored in `overlays/`.
+
+### Test Plan
+
+(**FIXME:** Mention that we will add unit tests)
+
+> **Note:** *Not required until targeted at a release.*
+>
+> Consider the following in developing a test plan for this enhancement:
+> - Will there be e2e and integration tests, in addition to unit tests?
+> - How will it be tested in isolation vs with other components?
+>
+> No need to outline all of the test cases, just the general strategy.  Anything
+> that would count as tricky in the implementation and anything particularly
+> challenging to test should be called out.
+>
+> All code is expected to have adequate tests (eventually with coverage
+> expectations).  Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+> when drafting this test plan.
+>
+> [testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+
+### Graduation Criteria
```

Comment:
My proposal for the **Graduation Criteria** section is the following:

- [ ] Implement the necessary functionality for this feature
- [ ] Write the appropriate unit tests
- [x] Add a user story in Kustomize's examples
- [ ] Extend Kustomize's glossary with a reference to components, as well as other places where overlays are mentioned
- [ ] Make the `TestCompositeDiamond` test, which was originally written to show that the composition feature was not supported, pass

Reply:
We might want to include `alpha -> beta` and `beta -> GA` sections here. Other KEPs seem to do this.

Other than that, the above criteria seem sufficient to me.